Add missing group change types

direc85 opened the pull request:
Attempt to add the missing group v2 change message types. This is literally "it compiles so I'm gonna push it" so it's possible, likely even, that I'm doing a thing or two wrong. I just looked how the parsing works for the existing message types and extrapolated from there. This is somewhere between a draft PR and a RFC...

I was expecting the handling to be absent here, but the parsing was pretty much complete, just missing the newer additions. So this means that this is just the tip of the iceberg and the client has to handle all those messages. Independently of this PR, I can start working on Whisperfish already. Other clients are most welcome to join the improve-a-thon!

rubdos commented on src/groups_v2/model.rs:
```diff
@@ -54,6 +54,34 @@ impl PartialEq for RequestingMember {
+#[derive(Derivative, Clone)]
+#[derivative(Debug)]
+pub struct BannedMember {
+    pub uuid: Uuid,
```
Should this be Aci?

direc85 replied:
I think so yes. Parts of the code path accept ServiceId when creating BannedMember (MemberBanned) in SA, but in all concrete cases I could only find use of UUIDs or ACIs.

rubdos commented on src/groups_v2/model.rs:
```diff
@@ -105,31 +133,31 @@ pub enum GroupChange {
     NewPendingMember(PendingMember),
     DeletePendingMember(Uuid),
     PromotePendingMember {
         uuid: Uuid,
```
Are pending members `Uuid`s, `ServiceId`s or `Aci`s? Maybe it's worth to comment the code whenever `Uuid` or `ServiceId` is used, to document *why* that specific form is used.

E.g., for invites, IIRC it's a `ServiceId`, because members can be invited based on their phone number (iirc).

direc85 replied:
Pending member is ServiceId indeed.

rubdos commented on src/groups_v2/model.rs:
```diff
@@ -105,31 +133,31 @@ pub enum GroupChange {
+    NewRequestingMember(RequestingMember),
+    DeleteRequestingMember(Uuid),
+    PromoteRequestingMember {
+        uuid: Uuid,
```
Same question

direc85 replied:
This is Aci, by SA sources. This makes sense, because if I request a group membership, I'll start the process directly with Aci, w.r.t. if someone is being invited, it may or may not be a known recipient.

rubdos commented:
@direc85 this is being tested, right? So maybe we should prioritize getting this merged? :-)

direc85 commented:
> this is being tested, right? So maybe we should prioritize getting this merged? :-)

Not yet actually. Two reasons: this is pretty much a somewhat-incorrect WIP (types need to be verified) and it's a nuisance to carry a dirty `Cargo.toml` between branches. I'm focusing on the Whisperfish part currently, but if someone wants to contribute, feel free to pick this up, I really don't mind :)

direc85 commented:
Worked this further; still likely not complete. Also, even less tested.

rubdos commented:
@direc85 how close to mergable is this?

direc85 commented:
> @direc85 this is being tested, right? So maybe we should prioritize getting this merged? :-)

Yes, I'm running this now nad have tested some group changes as well (add member, delete member).

> @direc85 how close to mergable is this?

Quite close; I'll still go through the types once (today-ish), and I'd like someone else have a look as well :)

rubdos commented on src/groups_v2/operations.rs:
```diff
@@ -84,6 +88,19 @@ impl GroupOperations {
+            .decrypt_service_id(bincode::deserialize(ciphertext)?)?
+        {
+            ServiceId::Pni(pni) => Ok(pni),
+            ServiceId::Aci(_aci) => Err(GroupDecodingError::NotPni),
```
It could be useful to already throw a `tracing::error` in here, since this is behavior that shouldn't happen and probably indicates a bug somewhere.

I could imagine that a pni-only field gets a separate encoding (e.g. without any service-id indicator) some day, and that we don't catch this. The same probably goes for ACI.

direc85 replied:
Done!

rubdos reviewed (approved):
This looks rather clean!

direc85 commented:
> TODO:
>     * `Member`: Uuid -> Aci, add Pni

Done, but adding Pni failed, because it's not in proto. Perhaps my assumption of lss `Member` == proto `DecryptedMember` is wrong, or we have to update protos or something? @rubdos any ideas?

>     * `PendingMember` Uuid -> Aci

Done.

>     * `BannedMember` Aci -> ServiceId

Done.

direc85 commented:
After reading even more SA code, I ended up changing more existing types (mostly Uuid &rarr; Aci).  There was quite a lot of changes since the last ack, so a proper re-review is in order.

I also added sending Blocked message request response, but made the executive decision to ignore blocked GroupV1's. Not even sure if we have the blocked attribute; GroupV1 is sufficiently dead that this can be safely skipped. Whisperfish for example didn't support the feature at all.

rubdos reviewed (commented):
Ack on ignoring the GV1 things, those are not useful anymore anyway.

direc85 commented on src/groups_v2/model.rs:
```diff
@@ -27,30 +27,83 @@ pub struct Member {
+    {
+        // We have to go through String deserialization,
+        // because Aci does not implement Deserialize (duh).
+        let s = String::deserialize(d)?;
```
Perhaps, will check.

direc85 commented on src/account_manager.rs:
```diff
@@ -902,7 +902,6 @@ impl AccountManager {
     }
 }
 
-#[expect(clippy::result_large_err)]
```
Commit message says: `Rust 1.75 doesn't like these.`

direc85 replied:
I can drop this commit sure, but then we'll have to whisperfork lss again... Thoughts?

rubdos replied:
I think we'll have to bump Rust either case for some recent changes :-(

Maybe that's some work for this summer.
